codegen.ml

```diff
@@ -1,191 +1,239 @@
+			let add_formal (t, n) p = 
+				L.set_value_name n p;
+				let local = L.build_alloca (ltype_of_typ t) n builder in
+					ignore (L.build_store p local builder);
+					ignore (Hashtbl.add local_vars n local)
+
+			in
+			List.iter2 add_formal fdecl.sformals (Array.to_list (L.params the_function));
 
     (* Return the value for a variable or formal argument.
        Check local names first, then global names *)
-    let lookup n = try StringMap.find n local_vars
-                   with Not_found -> StringMap.find n global_vars
+    let lookup n = try Hashtbl.find local_vars n
+                   with Not_found -> 
+                    try StringMap.find n global_vars 
+                     with Not_found -> raise (Failure ("undeclared identifier " ^ n))
     in
 
-
-  (* Construct code for an expression; return its value *)
-  (* let rec expr builder ((_, e) : sexpr) = *)
-  let rec expr builder ((_, e) : sexpr) =
-    match e with
-    | SIlit i -> L.const_int i32_t i
-    | SBLit b -> L.const_int i1_t (if b then 1 else 0)
-    | SFLit i -> L.const_float float_t (float_of_string i)
-    (* null? | SNoexpr     -> L.const_int i32_t 0 *)
-    | SId s -> L.build_load (lookup s) s builder
-    (* Matrix | SMatrixLit (contents, rows, cols) -> *)
-    | SBinop (e1, op, e2) ->
-        let e1' = expr builder e1 and e2' = expr builder e2 in
-        (match op with
-        | A.Add -> L.build_add
-        | A.Sub -> L.build_sub
-        | A.Mult -> L.build_mul
-        | A.Div -> L.build_sdiv
-        | A.And -> L.build_and
-        | A.Or -> L.build_or
-        | A.Equal -> L.build_icmp L.Icmp.Eq
-        | A.Neq -> L.build_icmp L.Icmp.Ne
-        | A.Less -> L.build_icmp L.Icmp.Slt
-        | A.Leq -> L.build_icmp L.Icmp.Sle
-        | A.Greater -> L.build_icmp L.Icmp.Sgt
-        | A.Geq -> L.build_icmp L.Icmp.Sge)
-          e1' e2' "tmp" builder
-    (* Unary and Negate *)
-    (* Function call *)
-    | SFunc (f, args) ->
+    (* Construct code for an expression; return its value *)
+    let rec expr builder ((_, e) : sexpr) =
+      match e with
+      | SILit i -> L.const_int i32_t i
+      (* null? | SNoexpr     -> L.const_int i32_t 0 *)
+      | SId s -> L.build_load (lookup s) s builder
+      (* Matrix | SMatrixLit (contents, rows, cols) -> *)
+      | SBinop (e1, op, e2) ->
+          let e1' = expr builder e1 and e2' = expr builder e2 in
+          (match op with
+          | A.Add -> L.build_add
+          | A.Sub -> L.build_sub
+          | A.Mul -> L.build_mul
+          | A.Div -> L.build_sdiv
+          )
+            e1' e2' "tmp" builder
+      (* Unary and Negate *)
+      (* Function call *)
+      | SFunc ("print", [e]) 
+      | SFunc ("printb", [e]) ->
+	      L.build_call printf_func [| int_format_str ; (expr builder e) |]
+	    "printf" builder      
+      | SFunc ("printf", [e]) -> 
+	      L.build_call printf_func [| float_format_str ; (expr builder e) |]
+	    "printf" builder
+      (* | SFunc (f, args) ->
```

Comment:
Looks like we are hardcoding the built-in functions. Are we planning on changing it later?

Reply:
I think built-in functions have to be hardcoded

Reply:
Or we can create a .tar file to share with the TA